File: cli/registration.js
import http from 'node:http';
import { addPushRegistration } from './store.js';

function readJsonBody(request) {
  return new Promise((resolve, reject) => {
    let body = '';

    request.on('data', (chunk) => {
      body += chunk;
      if (body.length > 128 * 1024) {
        request.destroy();
        reject(new Error('Request body is too large.'));
      }
    });

    request.on('end', () => {
      try {
        resolve(body.trim() ? JSON.parse(body) : {});
      } catch (error) {
        reject(error);
      }
    });

    request.on('error', reject);
  });
}

function sendJson(response, statusCode, payload) {
  response.writeHead(statusCode, {
    'content-type': 'application/json; charset=utf-8',
    'cache-control': 'no-store'
  });
  response.end(JSON.stringify(payload));
}

export function createRegistrationServer({ subscriptionId, port, host, onRegister }) {
  let resolveRegistration;
  const registered = new Promise((resolve) => {
    resolveRegistration = resolve;
  });

  const server = http.createServer(async (request, response) => {
    const url = new URL(request.url ?? '/', `http://${request.headers.host ?? '127.0.0.1'}`);

    if (request.method === 'GET' && url.pathname === '/health') {
      sendJson(response, 200, { ok: true, subscriptionId });
      return;
    }

    if (request.method !== 'POST' || url.pathname !== '/register') {
      sendJson(response, 404, { ok: false, error: 'Not found.' });
      return;
    }

    try {
      const payload = await readJsonBody(request);
      if (payload.subscriptionId !== subscriptionId) {
        sendJson(response, 403, { ok: false, error: 'Subscription mismatch.' });
        return;
      }

      const token = String(payload.pushToken ?? payload.token ?? '');
      if (!token) {
        sendJson(response, 400, { ok: false, error: 'Missing push token.' });
        return;
      }

      const register = onRegister ?? ((registration) => addPushRegistration(subscriptionId, registration));
      const { subscription } = await register({
        provider: String(payload.provider ?? 'fcm'),
        token,
        platform: String(payload.platform ?? 'unknown'),
        registeredAt: new Date().toISOString()
      });

      sendJson(response, 200, {
        ok: true,
        subscriptionId,
        pushTokenCount: subscription.pushTokens.length
      });
      resolveRegistration(subscription);
    } catch (error) {
      sendJson(response, 400, { ok: false, error: error.message });
    }
  });

  return {
    registered,
    start() {
      return new Promise((resolve, reject) => {
        server.once('error', reject);
        server.listen(port, host, () => {
          server.off('error', reject);
          resolve(server.address());
        });
      });
    },
    close() {
      server.close();
    }
  };
}

File: rep/internal/registration/registration.go
package registration

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net"
	"net/http"
	"time"
)

type Registration struct {
	Provider     string `json:"provider"`
	Token        string `json:"token"`
	Platform     string `json:"platform"`
	RegisteredAt string `json:"registeredAt"`
}

type RegisterFunc func(Registration) error

type Server struct {
	SubscriptionID string
	Port           int
	Host           string
	OnRegister     RegisterFunc

	registered chan struct{}
	subscription Registration
	httpServer   *http.Server
	listener     net.Listener
}

func New(subscriptionID string, port int, host string, onRegister RegisterFunc) *Server {
	if onRegister == nil {
		onRegister = func(Registration) error { return nil }
	}
	return &Server{
		SubscriptionID: subscriptionID,
		Port:           port,
		Host:           host,
		OnRegister:     onRegister,
		registered:     make(chan struct{}),
	}
}

func (s *Server) Registered() <-chan Registration {
	ch := make(chan Registration)
	go func() {
		<-s.registered
		ch <- s.subscription
		close(ch)
	}()
	return ch
}

func sendJSON(w http.ResponseWriter, status int, payload interface{}) {
	w.Header().Set("Content-Type", "application/json; charset=utf-8")
	w.Header().Set("Cache-Control", "no-store")
	w.WriteHeader(status)
	_ = json.NewEncoder(w).Encode(payload)
}

func (s *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	path := r.URL.Path

	if r.Method == http.MethodGet && path == "/health" {
		sendJSON(w, http.StatusOK, map[string]interface{}{
			"ok":             true,
			"subscriptionId": s.SubscriptionID,
		})
		return
	}

	if r.Method != http.MethodPost || path != "/register" {
		sendJSON(w, http.StatusNotFound, map[string]interface{}{
			"ok":    false,
			"error": "Not found.",
		})
		return
	}

	body, err := io.ReadAll(io.LimitReader(r.Body, 128*1024))
	if err != nil {
		sendJSON(w, http.StatusBadRequest, map[string]interface{}{"ok": false, "error": err.Error()})
		return
	}

	var payload map[string]interface{}
	if len(body) > 0 {
		if err := json.Unmarshal(body, &payload); err != nil {
			sendJSON(w, http.StatusBadRequest, map[string]interface{}{"ok": false, "error": err.Error()})
			return
		}
	}

	subID, _ := payload["subscriptionId"].(string)
	if subID != s.SubscriptionID {
		sendJSON(w, http.StatusForbidden, map[string]interface{}{
			"ok":    false,
			"error": "Subscription mismatch.",
		})
		return
	}

	token := ""
	if v, ok := payload["pushToken"].(string); ok && v != "" {
		token = v
	} else if v, ok := payload["token"].(string); ok && v != "" {
		token = v
	}
	if token == "" {
		sendJSON(w, http.StatusBadRequest, map[string]interface{}{
			"ok":    false,
			"error": "Missing push token.",
		})
		return
	}

	provider := "fcm"
	if v, ok := payload["provider"].(string); ok && v != "" {
		provider = v
	}
	platform := "unknown"
	if v, ok := payload["platform"].(string); ok && v != "" {
		platform = v
	}

	reg := Registration{
		Provider:     provider,
		Token:        token,
		Platform:     platform,
		RegisteredAt: time.Now().UTC().Format(time.RFC3339Nano),
	}

	if err := s.OnRegister(reg); err != nil {
		sendJSON(w, http.StatusBadRequest, map[string]interface{}{"ok": false, "error": err.Error()})
		return
	}

	s.subscription = reg
	close(s.registered)

	sendJSON(w, http.StatusOK, map[string]interface{}{
		"ok":             true,
		"subscriptionId": s.SubscriptionID,
		"pushTokenCount": 1,
	})
}

func (s *Server) Start() error {
	addr := fmt.Sprintf("%s:%d", s.Host, s.Port)
	listener, err := net.Listen("tcp", addr)
	if err != nil {
		return err
	}
	s.listener = listener
	s.httpServer = &http.Server{Handler: s}
	go func() {
		_ = s.httpServer.Serve(listener)
	}()
	return nil
}

func (s *Server) Address() string {
	if s.listener == nil {
		return ""
	}
	return s.listener.Addr().String()
}

func (s *Server) Close() error {
	if s.httpServer == nil {
		return nil
	}
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()
	return s.httpServer.Shutdown(ctx)
}
